### DISCUSSION TITLE
Issue populating relations

### DISCUSSION THREAD

**Discussion by yogeshkattel** [mikro-orm/mikro-orm#5761] [General]
### Describe the bug

I hope you're all doing well.

I'm currently using MikroORM 6.2 and facing an issue while populating relations data. When I try to populate, for example, the profile of a user, sometimes it populates correctly, but other times it doesn't. It seems random. I have tried using different populate strategies, but I'm still facing the same issue. I'm new to MikroORM and would appreciate any help or guidance on why this might be happening and how to fix it.

Sometimes the profile field is populated correctly, but other times it doesn't even have field and sometimes it has id but .

Steps I have tried:
Using different populate strategies.
Ensuring that the relations are properly defined in the entities.
Verifying that the database contains the related data.
Despite these efforts, the issue persists. Could someone please help me understand why this might be happening and how to resolve it?

It is very urgent because i have to submit the project in 2 days.
Thanks in advance!

### Reproduction

import { Cascade, Collection, Entity, Enum, ManyToMany, OneToMany, OneToOne, PrimaryKey, Property } from "@mikro-orm/core";
import CommonEntity from "common/entities/common.entity";
import PrimaryEntity from "common/entities/primary.entity";
import  {v4} from 'uuid';
import { Profile } from "profile/entities/profile.entity";
import { Otp } from "src/core/otp/entities/otp.entity";
import { Role } from "src/core/roles/entities/role.entity";
import { EmployeeType } from "types/enums/employee.enum";
import { EmployeeDetail } from "./employeeDetails.entity";
import { ExpertsDetail } from "./expert.entity";
import { ForumComment } from "src/core/forum-comment/entities/forum-comment.entity";
import { Forum } from "src/core/forum/entities/forum.entity";

@Entity()
export class User  extends CommonEntity{
  @PrimaryKey({
    name: 'user_id',
    type: 'uuid',
  })
  userId: string = v4();

 
 // other fields herer

  @OneToOne(() => Profile, (profile) => profile.user, {owner: true, nullable: true})
  profile: Profile

  @OneToMany(() => Otp, (otp)=> otp.user, {nullable: true})
  otp = new Collection<Otp>(this)
  
  @ManyToMany(() => Role, (role) => role.users, {nullable: true, owner: true})
  roles = new Collection<Role>(this)

  @OneToOne(() => EmployeeDetail, {nullable:true})
  employeeDetails:  EmployeeDetail

  @OneToOne(() => ExpertsDetail, {nullable: true})
  expertDetail: ExpertsDetail

  @ManyToMany(() => ForumComment, (comment) => comment.mentionedUsers, {nullable: true, owner: true, mappedBy: 'user_mentions'})
  mentions: ForumComment

  @OneToMany(() => Forum, forum =>forum.user, {nullable: true,})
  forums = new Collection<Forum>(this)

}

import { ArrayType, Entity, Enum, ManyToOne, OneToMany, OneToOne, Property } from "@mikro-orm/core";
import PrimaryEntity from "common/entities/primary.entity";
import { Images } from "src/core/imageupload/entities/imageupload.entity";

import { User } from "user/entities/user.entity";
import { FarmersKyc } from "./farmerkyc.entity";
import { AgroFirmKyc } from "./agrokyc.enum";

@Entity()
export class Profile extends PrimaryEntity {
  // other fields here
  @OneToOne(() => User, user => user.profile, {nullable: false})
  user: User

  

}

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2 

### Node.js version

18.20.3

### Operating system

garuda os

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Entity definitioin is not a reproduction, you will need to provide more.

Sounds like you are using global context, which could indeed cause this.

https://mikro-orm.io/docs/identity-map#problem-2---unstable-response-of-api-endpoints

**Reply by yogeshkattel** [reply]
I fixed it by using Loaded and LoadedCollection type:
netoOne(() => Images, {nullable: true})
picture: Loaded<Images>

@OneToMany(() => SubCategory, subCategory => subCategory.category)
subCategories: LoadedCollection

**Comment by sonicgamer0721** [comment]
@B4nan Maybe {nullable: true} is the real reason. Recently I found a new problem like this:

```
// database: Postgresql
// MikroOrm: 6.2.9
// NodeJS: 20.15.0
@Entity()
export class Label{
  ...
  @ManyToOne(() => User, { nullable: true })
  unlockedByUser: User | null;
}
```
I use em.find(Label) to get all labels with `populate: ['unlockedByUser']` or `fields: ['unlockedByUser.name']`.
If there are some labels whose unlockedByUser=null, it throws an error: **Cannot read properties of null (reading 'id')**
If all labels' unlockedByUser !==null, then everything works fine.

**Reply by sonicgamer0721** [reply]
Sorry, my fault, I found the reason is orderBy, I will create a new issue to talk about this.

**Reply by yogeshkattel** [reply]
The simple fix for me was to use Loaded type in my entity like:
  @OneToOne(() => Images, {nullable: true})
  picture: Loaded<Images>

  @OneToMany(() => SubCategory, subCategory => subCategory.category)
  subCategories:   LoadedCollection<SubCategory>

**Reply by B4nan** [reply]
> The simple fix for me was to use Loaded type in my entity like:

This is wrong, you are just lying to the compiler, making strictly typed relations non-working. Also this can have zero runtime effect, it cannot fix errors like `Cannot read properties of null (reading 'id')`.

**Reply by yogeshkattel** [reply]
Can you help me do it the right way. I was not able to fix it so i just used it so the work could at least progres. It would be great if you could tell the right way to do so. I just have to return the data to user from api and i dont need to process so maybe its not an issue ?

**Reply by B4nan** [reply]
But what? You just sent this short snippet here as an answer to something completely unrelated, idk what you are struggling with, but this report was about some runtime error, not a type-level issue. There are basically no reproductions anywhere in here, even the OP just provided entity definitions without showing what they are doing with them.
